src/update_eh.cpp

```diff
@@ -123,53 +123,74 @@ bool fields_chunk::update_eh(field_type ft, bool skip_w_components) {
+
+        // lazily allocate any E/H fields that are needed (H==B initially)
+        if (i == 0 && f[ec][cmp] == f[dc][cmp] &&
+            (s->chi1inv[ec][d_ec] || have_f_minus_p || dsigw != NO_DIRECTION)) {
+          f[ec][cmp] = new realnum[gv.ntot()];
+          memcpy(f[ec][cmp], f[dc][cmp], gv.ntot() * sizeof(realnum));
+          allocated_eh = true;
+        }
 
-      // for solve_cw, when W exists we get W and E from special variables
-      if (f_w[ec][cmp] && skip_w_components) continue;
+        // lazily allocate W auxiliary field
+        if (i == 0 && !f_w[ec][cmp] && dsigw != NO_DIRECTION) {
+          f_w[ec][cmp] = new realnum[gv.ntot()];
+          memcpy(f_w[ec][cmp], f[ec][cmp], gv.ntot() * sizeof(realnum));
+          if (needs_W_notowned(ec)) allocated_eh = true; // communication needed
+        }
 
-      // save W field from this timestep in f_w_prev if needed by pols
-      if (needs_W_prev(ec)) {
-        if (!f_w_prev[ec][cmp]) f_w_prev[ec][cmp] = new realnum[gv.ntot()];
-        memcpy(f_w_prev[ec][cmp], f_w[ec][cmp] ? f_w[ec][cmp] : f[ec][cmp],
-               sizeof(realnum) * gv.ntot());
-      }
+        // for solve_cw, when W exists we get W and E from special variables
+        if (f_w[ec][cmp] && skip_w_components) continue;
 
-      if (f[ec][cmp] != f[dc][cmp])
-        STEP_UPDATE_EDHB(f[ec][cmp], ec, gv, gv.little_owned_corner(ec), gv.big_corner(),
-                         dmp[dc][cmp], dmp[dc_1][cmp], dmp[dc_2][cmp],
-                         s->chi1inv[ec][d_ec], dmp[dc_1][cmp] ? s->chi1inv[ec][d_1] : NULL,
-                         dmp[dc_2][cmp] ? s->chi1inv[ec][d_2] : NULL, s_ec, s_1, s_2, s->chi2[ec],
-                         s->chi3[ec], f_w[ec][cmp], dsigw, s->sig[dsigw], s->kap[dsigw]);
+        // save W field from this timestep in f_w_prev if needed by pols
+        if (i == 0 && needs_W_prev(ec)) {
+          if (!f_w_prev[ec][cmp]) f_w_prev[ec][cmp] = new realnum[gv.ntot()];
+          memcpy(f_w_prev[ec][cmp], f_w[ec][cmp] ? f_w[ec][cmp] : f[ec][cmp],
+                 sizeof(realnum) * gv.ntot());
+        }
+
+        // specify tile's start and end ivecs
+        ivec ts = gvs[i].little_owned_corner(ec);
+        ivec te = gvs[i].big_corner();
+
+        if (s->chi1inv[ec][d_1] == NULL && s->chi1inv[ec][d_2] == NULL) {
+          ts = gv.little_owned_corner(ec);
+          te = gv.big_corner();
+        }
+
+        if (f[ec][cmp] != f[dc][cmp])
+          STEP_UPDATE_EDHB(f[ec][cmp], ec, gv, ts, te,
+                           dmp[dc][cmp], dmp[dc_1][cmp], dmp[dc_2][cmp],
+                           s->chi1inv[ec][d_ec], dmp[dc_1][cmp] ? s->chi1inv[ec][d_1] : NULL,
+                           dmp[dc_2][cmp] ? s->chi1inv[ec][d_2] : NULL, s_ec, s_1, s_2, s->chi2[ec],
+                           s->chi3[ec], f_w[ec][cmp], dsigw, s->sig[dsigw], s->kap[dsigw]);
+
+        if (s->chi1inv[ec][d_1] == NULL && s->chi1inv[ec][d_2] == NULL && cmp == 1 - is_real)
```

Comment:
I would add this logic when you set up the `gvs` list in the first place — if the material is isotropic, then skip the whole subdivision of `gv` and set `gvs` to be a list of a single "tile" (the whole `gv`).

Reply:
This would require you to have a separate `gvs` list for the `step_update_edhb`, which be a good idea anyway since minimum tile size might be different.

Reply:
You'll want to set up these `gvs` lists whenever `changed_materials` is true, so you might as well call the routine to set them up right in `step_*` before they are needed.   i.e.
```
if (changed_materials)
    gvs_edhb = ....;
```